### PR TITLE
Runtime schema declaration

### PR DESCRIPTION
**Pull request by crwen**
support runtime schema

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -69,6 +69,23 @@ jobs:
           command: fmt
           args: -- --check
 
+  exmaples:
```
Comment (KKould) [inline .github/workflows/ci.yml:72]: already exists: https://github.com/tonbo-io/tonbo/blob/main/.github/workflows/ci.yml#L72

File: Cargo.toml [hunk 2]
```diff
@@ -49,21 +49,23 @@ path = "benches/criterion/writes.rs"
...
 async-stream = "0.3"
 async-trait = { version = "0.1", optional = true }
 bytes = { version = "1.7", optional = true }
 crc32fast = "1"
 crossbeam-skiplist = "0.1"
-datafusion = { version = "41", optional = true }
+datafusion = { version = "42", optional = true }
 flume = { version = "0.11", features = ["async"] }
+fusio = { git = "https://github.com/tonbo-io/fusio.git", package = "fusio", rev = "317b1b0621b297f52145b41b90506632f2dc7a1d", features = ["tokio", "dyn"] }
+fusio-parquet = { git = "https://github.com/tonbo-io/fusio.git", package = "fusio-parquet", rev = "317b1b0621b297f52145b41b90506632f2dc7a1d" }
```
Comment (KKould) [inline Cargo.toml:62]: Currently the latest is: ab6a8073b966f9c7d7ca61c32646b7d45dbc4f31

File: src/inmem/immutable.rs [hunk 3]
```diff
@@ -53,16 +54,25 @@ where
...
+        SkipMap<Timestamped<<A::Record as Record>::Key>, Option<A::Record>>,
+        &RecordInstance,
+    )> for Immutable<A>
 where
     A: ArrowArrays,
     A::Record: Send,
 {
-    fn from(mutable: SkipMap<Timestamped<<A::Record as Record>::Key>, Option<A::Record>>) -> Self {
+    fn from(
+        mutable: (
```
Comment (KKould) [inline src/inmem/immutable.rs:67]: naming variables separately is more readable

```rust
(mutable, instance): (
  SkipMap<Timestamped<<A::Record as Record>::Key>, Option<A::Record>>,
  &RecordInstance,
)
```

File: src/inmem/mutable.rs [hunk 4]
```diff
@@ -348,4 +341,55 @@ mod tests {
...
+            .await
+            .unwrap();
+
+        {
+            let mut scan = mutable.scan((Bound::Unbounded, Bound::Unbounded), 0_u32.into());
+            let entry = scan.next().unwrap();
+            assert_eq!(
+                entry.key(),
+                &Timestamped::new(
+                    Column::new(Datatype::Int8, "age".to_string(), Arc::new(1_i8), false),
```
Comment (KKould) [inline src/inmem/mutable.rs:388]: why is it just the value of one column in a record? Before, each mutable entry should be a record?

File: src/lib.rs [hunk 5]
```diff
@@ -170,12 +179,50 @@ where
...
+            base_fs
+                .create_dir_all(&option.version_log_dir_path())
+                .await
+                .map_err(DbError::Fusio)?;
+        }
+
+        let instance =
+            RecordInstance::Runtime(DynRecord::empty_record(column_descs, primary_index));
+
+        let db = Self::build(option, executor, instance, manager).await?;
```
Comment (KKould) [inline src/lib.rs:220]: It seems that the variable `db` can be removed

File: src/lib.rs [hunk 6]
```diff
@@ -576,12 +670,13 @@ where
...
         }
-        let mut fixed_projection = vec![0, 1, R::primary_key_index()];
+        let primary_key_index = self.schema.record_instance.primary_key_index::<R>();
+        let mut fixed_projection = vec![0, 1, primary_key_index];
         fixed_projection.append(&mut projection);
         fixed_projection.dedup();
 
         let mask = ProjectionMask::roots(
-            &arrow_to_parquet_schema(R::arrow_schema()).unwrap(),
+            &arrow_to_parquet_schema(&self.schema.record_instance.arrow_schema::<R>()).unwrap(),
```
Comment (KKould) [inline src/lib.rs:679]: there is no need to call this method `self.schema.record_instance.arrow_schema::<R>()`, use `primary_key_index` directly

File: src/record/mod.rs [hunk 7]
```diff
@@ -9,13 +10,57 @@ use arrow::{array::RecordBatch, datatypes::Schema};
...
+        R: Record,
+    {
+        match self {
+            RecordInstance::Normal => R::arrow_schema().clone(),
+            RecordInstance::Runtime(record) => record.arrow_schema(),
+        }
+    }
+
+    pub(crate) fn size(&self) -> usize {
+        todo!()
```
Comment (KKould) [inline src/record/mod.rs:60]: the lack of `size()` implementation should cause the data to be unable to trigger Compaction (the threshold will not be reached)

File: src/record/mod.rs [hunk 8]
```diff
@@ -9,13 +10,57 @@ use arrow::{array::RecordBatch, datatypes::Schema};
...
+        }
+    }
+
+    pub(crate) fn primary_key_path<R>(&self) -> (ColumnPath, Vec<SortingColumn>)
+    where
+        R: Record,
+    {
+        match self {
+            RecordInstance::Normal => R::primary_key_path(),
+            RecordInstance::Runtime(_) => todo!(),
```
Comment (KKould) [inline src/record/mod.rs:45]: missing implementation

File: src/record/mod.rs (hunk 7 above)
Comment (crwen) [reply src/record/mod.rs:60]: This method is not used, I will delete it

File: src/record/runtime/array.rs [hunk 9]
```diff
@@ -0,0 +1,373 @@
...
+            return Some(None);
+        }
+
+        let mut columns = vec![];
+        for (idx, col) in self.columns.iter().enumerate() {
+            if projection_mask.leaf_included(idx + 2) && !col.is_nullable {
+                let datatype = col.datatype;
+                let name = col.name.to_string();
+                let value = match datatype {
+                    Datatype::Int8 => {
```
Comment (KKould) [inline src/record/runtime/array.rs:84]: I think macros can be used to simplify this to simply implement most types.

File: src/record/runtime/record.rs [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+
+impl Record for DynRecord {
+    type Columns = DynRecordImmutableArrays;
+
+    type Key = Column;
+
+    type Ref<'r> = DynRecordRef<'r>;
+
+    fn primary_key_index() -> usize {
+        unreachable!("This method is not used.")
```
Comment (KKould) [inline src/record/runtime/record.rs:155]: can this method be used to implement another trait such as `StaticRecord` instead of `unreachable!` ?

**Review by ethe — APPROVED** [review]

**Pull request merged**